### ISSUE TITLE
applications not linked in a way finder can find them

### ISSUE THREAD

**Issue by mknudsen** [LnL7/nix-darwin#1079]
I have a bunch of apps installed through nix-darwin and I can launch them from command line, i.e. [Zoom](https://search.nixos.org/packages?channel=unstable&from=0&size=50&buckets=%7B%22package_attr_set%22%3A%5B%5D%2C%22package_license_set%22%3A%5B%5D%2C%22package_maintainers_set%22%3A%5B%5D%2C%22package_platforms%22%3A%5B%22aarch64-darwin%22%5D%7D&sort=relevance&type=packages&query=zoom) with `zoom`.

I also see them in `/run/current-system/Applications/`. I however cannot find them through finder. How do I make macOS aware or those apps?

`darwin-rebuild check` tells me everything is ok.

Here are my software versions

```
darwin-version 
24.11pre674705.b833ff01a0d6+darwin4

Darwin myhostname 24.0.0 Darwin Kernel Version 24.0.0: Mon Aug 12 20:49:48 PDT 2024; root:xnu-11215.1.10~2/RELEASE_ARM64_T8103 arm64
```

**Comment by emilazy** [comment]
Aliases also have significant issues; see https://github.com/nix-community/home-manager/issues/1341 for discussion. Probably copying is the best solution.

**Comment by antoineco** [comment]
@emilazy are you referring to these two points?

> Aliases won't be categorized under "Applications" [by Spotlight]

> Aliases are kind of awful to work with - you can't seem to create them from the CLI without pinging finder via AppleScript / osascript

**Comment by emilazy** [comment]
Yes, though also making aliases through Finder is a hack with its own set of problems so we’d want to use a native code solution to accomplish it even if we went that route. The Spotlight problem is pretty serious though. If there was something other than copying that worked well I expect Homebrew Cask would have already adopted it.

**Comment by reckenrode** [comment]
Aliases can be created programatically (e.g., https://github.com/reckenrode/mkalias). Finder should not be scripted. It’s [undermines TCC][1]. macOS 15 also [requires][2] explicit approve Finder of scripting in Security & Privacy, making the experience even worse (probably to discourage scripting of Finder because it’s effectively equivalent to having Full Disk Access).

(I’m also pretty sure aliases share similar limitations to the trampoline method with references to the store in the dock and services not being able until an aliased application has been launched, but it’s been so long since I looked into it that I don’t remember for sure.)

[1]: https://labs.sentinelone.com/bypassing-macos-tcc-user-privacy-protections-by-accident-and-design/
[2]: https://developer.apple.com/documentation/macos-release-notes/macos-15-release-notes#New-Features

**Comment by wellerbp** [comment]
Is your concern simply finding them in spotlight?

**Comment by mknudsen** [comment]
> Is your concern simply finding them in spotlight?

Yes, cmd + space happens to be the way I start most my apps.

**Comment by fbettag** [comment]
Same goes for Alfred, can't find it if it isn't in Applications.

**Comment by Enzime** [comment]
Raycast shows my nix-darwin (`/Applications/Nix Apps`) and home-manager (`~/Applications/Home Manager Apps`) apps correctly 👍

**Comment by wellerbp** [comment]
> > Is your concern simply finding them in spotlight?
> 
> Yes, cmd + space happens to be the way I start most my apps.

There are a few ways to accomplish this, at least of which I am aware,  but each has a downside or limitation. The best option is to create a symlink in your `Applications` that will have all installed applications, cli tools, etc. appear as `Other`.

Try entering this block into your configuration:

```
in
        pkgs.lib.mkForce ''
          # Set up applications.
          echo "setting up /Applications..." >&2
          rm -rf /Applications/Nix\ Apps
          mkdir -p /Applications/Nix\ Apps
          find ${env}/Applications -maxdepth 1 -type l -exec readlink '{}' + |
          while read src; do
            app_name=$(basename "$src")
            echo "copying $src" >&2
            ${pkgs.mkalias}/bin/mkalias "$src" "/Applications/Nix Apps/$app_name"
          done
          find ${env}/Applications -maxdepth 1 -type l -exec readlink '{}' + |
        '';
 ```

**Comment by dwt** [comment]
Out of interest, does anybody know the downsides of using directory hard links or clonefile to try to solve these issues?

**Comment by reckenrode** [comment]
Directory hard links aren’t supported on APFS.

**Comment by reckenrode** [comment]
I’ll reiterate my position that copying is the least bad solution. The only limitation is that bundles with self-referencing, absolute paths need to have them converted into bundle-relative paths.

**Comment by dwt** [comment]
Well, [I've tried to to get a draft going how this could look](https://github.com/LnL7/nix-darwin/pull/1396). Since I can't test this (yet) as I don't want to switch my main system to nix-darwin master, this is likely to not work and kill your system if you try. Some caution may be advised.

Still I would love some feedback and ideas on how to improve it!

**Comment by emilazy** [comment]
There’s really not that much difference between `master` and the release branch right now, FWIW. (Except that `master` expects you to use `nixpkgs-unstable`, but the main risk there is just “some packages might have broken builds depending on the phase of the moon”.)

**Comment by Samasaur1** [comment]
Closed by #1396
